### DISCUSSION TITLE
502 Bad Gateway if a container restarts

### DISCUSSION THREAD

**Discussion by yurividal** [ZoeyVid/NPMplus#1237] [Q&A] [answered]
I have my npmplus running on a docker-network. I have other containers join the same docker network, to be proxied.
For example:
For my vaultwarden container, i proxy htttp:vaultwarden:80 and it works. But, if i restart the vaultwarden container, npmPlus stops proxying to that container. I start getting 502 Bad gateway. I have to restart npmPlus, and it starts working again.

This is a problem everytime watchtower updates any of my containers. 

This didn't happen with npm

**Comment by Zoey2936** [comment, marked as answer]
1. watchtower is not a good solution to update anything
2. npmplus is written for network_mode host
3. If you use docker networks, then use static ips instead of hostnames

**Reply by yurividal** [reply]
Thank you for your response @Zoey2936 
I'll have to migrate all my containers to static docker IPs, and use static docker overlay IPs.
But, i'm curious to know... any ideas why this happens with npmPlus and doesn't happen with npm or vanilla nginx? Is this a DNS timeout?

**Reply by Zoey2936** [reply]
not sure, NPM uses openresty, NPMplus uses freenginx with this dns patch: https://raw.githubusercontent.com/openresty/openresty/master/patches/nginx-1.27.1-resolver_conf_parsing.patch
to support: this config line:
https://github.com/ZoeyVid/NPMplus/blob/7c3d9947e98afb443b4678b9870a11e121be4cd6/rootfs/usr/local/nginx/conf/nginx.conf#L63
local=on means that we use the resolvers set in the /etc/resolv.conf file inside the container

But in any case static IPs is the better solution because dns can always fail

**Reply by Zoey2936** [reply]
for intrest what is written in the docker logs of npmplus?

**Reply by parasiteoflife** [reply]
I'm having the same problem, what I think it's happening is: first time NPMplus is started -and also maybe when the proxy host is added I haven't tested- and we are using hostnames instead of direct IPs like OP, NPMplus resolves the hostname to an IP and caches the IP for the rest of the session then when a container is restarted (I don't know if every time but at least the times I have noticed) their IP changes thus the cached IP NPMplus has is no longer valid hence the error that is returned. That's why restarting NPMplus works, because it retrieves the IPs again.

This might also pose security risks:

1. Add a proxy host using hostname private-A not publicly available (using access lists)
2. Add a proxy host using hostname public-B publicly available
3. Restart container public-B, it takes a different IP
4. Restart container private-A, it takes the -now available- IP public-B was using before being restarted
5. Because NPMplus uses the cached IP when we access the proxy public-B it will serve private-A instead breaking the security that was setup.

**Reply by Zoey2936** [reply]
yes this could happen and the fix is still: use static ips

**Reply by parasiteoflife** [reply]
> yes this could happen and the fix is still: use static ips

but that is almost impossible to do when using docker networks. When we spin up a compose file almost never the container's IP is setup manually, docker manages that, even using docker networks is a security practice to not expose ports and have the facility to use the containers hostname to use with any reverse proxy or even other containers.

You are saying we should start setting up docker IPs manually for all the services? Because I tried setting the IP of some (not all) containers manually in the past but it didn't worked well, if I had a compose file that used 4 services, all of them using the same network, when I tried just setting up only the IP of one of the services manually it broke constantly because docker wanted to set that IP for one of the other 3 containers when docker restarted or the compose file was re-deployed and other shenanigans that also happened too, I even tried to use the last IP (.254) but that also failed. I used to do that instead of using hostnames, like to connect other containers to my ntfy instance but because it was so fragile I started using hostnames and problem solved.

So please add support for this!

**Reply by Zoey2936** [reply]
If you tell me how to add support for this I will do, until then use static ips
